pallets/unicks/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+//! # Pallet storing unique nickname <-> DID links for user-friendly DID
+//! nicknames.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod types;
+mod utils;
+
+#[cfg(test)]
+mod mock;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use codec::FullCodec;
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{Currency, ReservableCurrency, StorageVersion},
+		Blake2_128Concat,
+	};
+	use frame_system::pallet_prelude::*;
+	use sp_std::{fmt::Debug, vec::Vec};
+
+	use kilt_support::{deposit::Deposit, traits::CallSources};
+
+	use crate::types::UnickOwnership;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	pub type UnickInput<T> = BoundedVec<u8, <T as Config>::MaxUnickLength>;
+
+	pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+	pub(crate) type BalanceOf<T> = <<T as Config>::Currency as Currency<AccountIdOf<T>>>::Balance;
+	pub(crate) type BlockNumberFor<T> = <T as frame_system::Config>::BlockNumber;
+	pub(crate) type CurrencyOf<T> = <T as Config>::Currency;
+	pub(crate) type UnickOwnerOf<T> = <T as Config>::UnickOwner;
+	pub(crate) type UnickOf<T> = <T as Config>::Unick;
+	pub(crate) type UnickOwnershipOf<T> =
+		UnickOwnership<UnickOwnerOf<T>, Deposit<AccountIdOf<T>, BalanceOf<T>>, BlockNumberFor<T>>;
+
+	// Internal type used to differentiate the origin of a `release` call.
+	enum UnickReleaseCaller<'a, 'b, T: Config> {
+		// The origin is an account ID that should match the deposit payer.
+		DepositPayer(&'a AccountIdOf<T>),
+		// The origin is a DID that should match the unick owner.
+		UnickOwner(&'b UnickOwnerOf<T>),
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	// Unick -> owner
+	#[pallet::storage]
+	#[pallet::getter(fn owner)]
+	pub type Owner<T> = StorageMap<_, Blake2_128Concat, UnickOf<T>, UnickOwnershipOf<T>>;
+
+	// Owner -> Unick
+	#[pallet::storage]
+	#[pallet::getter(fn unicks)]
+	pub type Unicks<T> = StorageMap<_, Twox64Concat, UnickOwnerOf<T>, UnickOf<T>>;
+
+	// Unick -> ()
+	#[pallet::storage]
+	#[pallet::getter(fn is_blacklisted)]
+	pub type Blacklist<T> = StorageMap<_, Blake2_128Concat, UnickOf<T>, ()>;
```

Comment:
We only need the `Unicks` store for the front end. 😕

Reply:
When exposing to the SDK, I guess we would also need to fetch the currently blacklisted one (to throw errors before even submitting the tx), and probably also the unick owned by a DID, cause why not?

Reply:
no way around the convenience storage. 🤷🏻